### ISSUE TITLE
Support for updating chain parameters using xfield

### ISSUE THREAD

**Issue by azuchi** [chaintope/tapyrus-signer#163]
Tapyrus block headers currently support the following xfield types:

* 1: Aggregate public key.
* 2: Max block size

Add to Tapyrus Signer the ability to generate new blocks by specifying these xfield types. The basic flow is as follows:

1. First, for each Signer node, register the xfield change by specifying the xfield type and value to be changed and the block height to include it (these data should be persistent). It is necessary to specify a block height with sufficient margin to allow other nodes to finish their operations.
2. The master node creating the candidate block to apply the change if it has an outstanding xfield type change and its block height is greater than or equal to the block to be created (specify them in the tapyrus core RPC).
3. When a node receives a candidate block from the master node, if the candidate block contains a change of xfield type, it checks to see if the change has been registered on its own node.
   * If so, it approves it and does nothing.
   * If not, it treats the candidate block as invalid.
4. If a signature can be generated for the candidate block for the threshold and the block is complete, the block that updates the xfield type is successfully generated.

In addition to above, add a CLI command to delete registered xfield information (to delete data registered in error).

**Comment by Naviabheeman** [comment]
PR #167 adds federation command. Some questions at this point to discuss are:
1. Where to write the federation change file? can we keep it as a fixed default directory and document it or make it configurable?
2. All register and unregister commands change the same file or create multiple files ( which is the current behaviour)?

**Comment by azuchi** [comment]
> Where to write the federation change file? can we keep it as a fixed default directory and document it or make it configurable?

Let $HOME/.tapyrus-signer be the base data directory. The data directory can be changed using the configuration option `--datadir`. This setting can be specified as a command line argument or in a toml file like `--log-file` option. 

> All register and unregister commands change the same file or create multiple files ( which is the current behaviour)?

I think it's fine to use the same file. Does it need to be separated?

**Comment by Naviabheeman** [comment]
@azuchi 
Signer code has logic to support multiple federations as it is.  The master node does `getnewblock`, `verify_block` and  `add_aggregated_public_key_if_needed` for each block. When `federations.toml` is extended with multiple federation definitions this code can make the aggregated pub key change in the block header before signing the block. This is not verified by other signers though.

I'm implementing 'federation' command like this:
- write all updates to one file, `federationchange.dat`. 
- The signer node would monitor this file and sync the changes to an in memory list. 
- when the block height is reached, it adds the xfield to the block. 
- remove that completed entry from federationchange.dat if the xfield change was successful in the that round
- When the block fails, the change would be retried by the next master node. 

1. If we add `threshold` and 'node-vss` to this command, it would be better to write to  `federations.toml`. So change I'm aggregated-pub-key goes to  `federations.toml` and change in max-block-size  goes to `federationchange.dat`.  Then running one command on all signer nodes will be enough to change any xfield.
2. we can simplify the same strategy to use just `federations.toml` for all xfield changes. we can extend the file schema to support all xfields. Then federation command can write to one file and we will need only one command for all xfield changes. 
3. we can use `federations.toml` as a reference for the federation command. The command would verify if the federation details, `threshold` and 'node-vss` are added  to `federations.toml`. Check their validity and then only add aggregated pub key to  `federationchange.dat`. In this case we the federation to update `federations.toml` before aggregate-pub-key change.  So there is more housekeeping. 
Which of these would be preferable?

**Comment by azuchi** [comment]
> Signer code has logic to support multiple federations as it is. The master node does getnewblock, verify_block and add_aggregated_public_key_if_needed for each block.

Yes. The master node switches in a round-robin fashion.  The minimum number of federations is 3. 
For now, we assume that the federation is composed of members in a trusting relationship. In other words, updating the aggregate public key(updating the federation members) must be agreed upon in the real world.

> I'm treating this file as a list of pending changes.

If a command is executed multiple times and multiple xfield changes occur, are they all recorded?

> we can simplify all of the above and use just federations.toml for all xfield changes.

In this case, you mean  all changes are directly editing the file?

**Comment by Naviabheeman** [comment]
> If a command is executed multiple times and multiple xfield changes occur, are they all recorded?

The file is guarded by a lock to protect it within a process. And I added a test to run two commands in parallel.  
It does not cause any failure yet. This conversation has given me an idea to extend the test i.e to repeat the command with a new block height and verify that all writes were recorded. 
Otherwise we need to go back to multiple file outputs for each command execution described in [this comment](https://github.com/chaintope/tapyrus-signer/issues/163#issuecomment-1725543342)


> In this case, you mean all changes are directly editing the file?

Yes. I mean edit the file directly.  I would also like to add a verification field like signature for each change. If the whole federation was involved in making the decision to change the parameter the signature can be generated for all xfields like aggregated public key.

**Comment by azuchi** [comment]
> Yes. I mean edit the file directly.

I see. Let's implement this(method 3) instead of the command first.  Also, I don't think you need to monitor the file, just load it at startup.

**Comment by Naviabheeman** [comment]
> I see. Let's implement this(method 3) instead of the command first. 

Do you mean with a verifiable signature or without it? Probably this:

```
[[xfields]]
{
block_height = <bock_height>
<xfield_name> = <xfield_value>
signature = <signature>
}
```

>Also, I don't think you need to monitor the file, just load it at startup.

Sure.

**Comment by azuchi** [comment]
> Do you mean with a verifiable signature or without it? Probably this:

If you add a signature;

* Is this required not only for aggregate public key updates, but also for updates of other xfield types(like max block size)?
* What message and public key is this signature for?

**Comment by Naviabheeman** [comment]
> * Is this required not only for aggregate public key updates, but also for updates of other xfield types(like max block size)?

I think we need the signature for all xfield changes other than aggregated-pub key. In the case of aggregated pub key change, the node vss needs to be generated with the cooperation of all or majority of signers. Other xfields like maxblocks size can be added by anyone with access to the file. we can have a threshold signature, similar to block proof. i.e. sign the `hash(serialized_xfield)` using the distributed signature scheme. This will ensure that all signers are aware of the change in xfield. 

So the signer implementation would:
- read `federations.toml`
- verify the signature in each xfield change using the aggregated pub key of the federation at that block height.
- while generating the block with xfield, master node can add the xfield to block header.
- other nodes participating in the block generation can verify the xfield in their federation.toml, verify its signature and then provide the partial signature to the block.


> * What message and public key is this signature for?

sign the `hash(serialized_xfield)` using the distributed signature scheme. and verify using aggregated-pub key at that block-height.

**Comment by azuchi** [comment]
I see. Let's add a signature field. 

> verify the signature in each xfield change using the aggregated pub key of the federation at that block height.

for example, when updating a new aggregate public key with a block height of 100, signature public key is the aggregate public key before the update, right?

**Comment by Naviabheeman** [comment]
Earlier I meant that we don't need a signature for aggregated pub key change as their node vss can ensure it is secure. But on the other hand it's not uniform. So I think yes, we can add a signature to aggregated pub key change too. 

> for example, when updating a new aggregate public key with a block height of 100, signature public key is the aggregate public key before the update, right?

Yes, the signature pub key is the aggregate public key before the update.

**Comment by Naviabheeman** [comment]
#169 and #170 add the needed support. 

TODO:
Add more tests - multiple federations with multiple federation-changes

Add a maximum retries entry to federation change struct to limit the number of times an xfield change must be retried before it is discarded. Otherwise the block generation may be compromised. It can be equal to the threshold by default.

Add a criteria to each xfield to verify its validity. 
i.e aggregated pub key - compressed and validated in secp256k1
max-block-size - have minimum and maximum limit matching Tapyrus-core
